### PR TITLE
Melt relational primops

### PR DESCRIPTION
**Pull request by amosr**
! @tranma @jystic 
/jury approved @jystic

### CODE REVIEW

**Comment by amosr** [comment]
#472

File: icicle-data/src/Icicle/Common/Base.hs [hunk 1]
```diff
@@ -90,8 +90,8 @@ data BaseValue
  | VPair     !BaseValue !BaseValue
  | VLeft     !BaseValue
  | VRight    !BaseValue
- | VSome     !BaseValue
  | VNone
+ | VSome     !BaseValue
```
Comment (amosr) [inline icicle-data/src/Icicle/Common/Base.hs]: had to reorder these so `None < Some 0`

**Comment by jystic** [comment]
From the failed build:
```
=== prop_read_or_drop_entity from test/Icicle/Test/Sea/Zebra.hs:97 ===
--- many lines snipped ---
error: #error Failed during codegen (seaOfXValue: buffer elements should be converted to pushes by convertValues : (VBuf [VString \"bart\"],BufT 1 StringT)..)\n")
```

File: icicle-compiler/test/Icicle/Test/Arbitrary/Core.hs [hunk 2]
```diff
@@ -193,16 +205,56 @@ instance Arbitrary ValType where
...
+         [ IntT
+         , UnitT
+         , BoolT
+         , TimeT
+         , StringT ]
+         [-- ArrayT  <$> genPrimType
+         --, BufT    <$> (getPositive <$> arbitrary) <*> genPrimType
+           PairT   <$> genOrdValType <*> genOrdValType
+         , SumT    <$> genOrdValType <*> genOrdValType
+         -- , MapT    <$> genPrimType <*> genPrimType
```
Comment (jystic) [inline icicle-compiler/test/Icicle/Test/Arbitrary/Core.hs]: Should commented out things get removed?

File: icicle-compiler/test/Icicle/Test/Arbitrary/Corpus.hs [hunk 3]
```diff
@@ -59,11 +59,47 @@ queries =
...
+   , "           end"
+   , "         | False ->"
+   , "           case is_homer"
+   , "           | True -> (map_insert location severity map_create, Some time)"
+   , "           | False -> (map_create, Some time)"
+   , "           end"
+   , "         end"
+   , "       end"
+   , "     end"
+   , "~> (group fold (k, v) = fst x ~> min v)"
```
Comment (jystic) [inline icicle-compiler/test/Icicle/Test/Arbitrary/Corpus.hs]: Wow, @markhibberd would be proud 💯

**Comment by jystic** [comment]
👍 on 🍏

**Comment by tranma** [comment]
I'm not sure nested `Buf`s can be a thing? Group in group will allow nested arrays but nested bufs probably can't exist.

:1st_place_medal: otherwise!

**Comment by amosr** [comment]
> I'm not sure nested Bufs can be a thing? Group in group will allow nested arrays but nested bufs probably can't exist.

What about Buf (Array _)?

**Comment by tranma** [comment]
I think `Buf (Array t)` might be generated with the right kind of group

File: icicle-data/src/Icicle/Data/Time.hs [hunk 4]
```diff
@@ -112,7 +112,7 @@ localMinute :: Time -> Int
 localMinute x = timeOfDay x ^. Thyme._todMin
 
 localSecond :: Time -> Int
-localSecond x = timeOfDay x ^. Thyme._todHour
+localSecond x = truncate (timeOfDay x ^. Thyme._todSec)
```
Comment (tranma) [inline icicle-data/src/Icicle/Data/Time.hs]: wow 🇦🇺

Comment (jystic) [inline icicle-data/src/Icicle/Data/Time.hs]: ⌛️

**Comment by amosr** [comment]
going to merge this hellhole

**Comment by jystic** [comment]
🍾

**Pull request merged**